Automatically download manifest from server

christianbundy opened the pull request:
Manifests are hard and the server is happy to provide one for us.

This simplifies ssb-client to remove the `manifest` option, using muxrpc's [manifest bootstrap](https://github.com/ssbc/muxrpc/pull/56) instead of asking the user or storing one at `~/.ssb/manifest.json`.

I don't think we want to support any use-case where the client has a different manifest than the server, please correct me if I'm wrong.

christianbundy commented on client.js:
```diff
@@ -53,21 +51,36 @@ module.exports = function (opts, cb) {
+    // HACK: This refers to the *remote* feed ID, not our local ID, and we don't
+    // actually have a way of determining the feed type. This is fine when all
+    // feeds are ed25519, but will cause problems in the future.
+    const remoteId = `@${stream.remote.toString('base64')}.ed25519`
```
Would love another pair of eyes to ensure that this looks right.

dominictarr replied:
good comment!

christianbundy commented on client.js:
```diff
@@ -53,21 +51,36 @@ module.exports = function (opts, cb) {
+      // same when connecting to a remote server.
+      if (sbot.id == null) {
+        sbot.id = opts.keys.id
+      }
```
Would also love for someone to double-check this change and its documentation to make sure they're right.

christianbundy replied:
(This was previously `sbot.id = remoteId`, which I think was wrong (?).)

dominictarr replied:
why do you think that is wrong? it's useful to know who you are talking too... I feel there is probably some code somewhere that uses that.

christianbundy replied:
The code was added in https://github.com/ssbc/ssb-client/pull/50/commits/7906cb364ceff0c558d8d69c74146888ff8cd6fb and I think either the comments are backward, the logic is backward, or I'm misreading something.

That commit defines `id` as the remote ID, comments `id === config.keys.id`, and puts the remote in `sbot.id`, which I thought was supposed to be the same as `sbot.whoami().id` (the local ID). Could you help be me less confused? I wouldn't be surprised if I'm just reading it wrong.

dominictarr replied:
there is the server's id, the clients id. whoami returns the servers id. on a local instance of sbot, sbot.id is the server id... so `rpc_sbot.id` ought to be the remote id? that would make the it possible for the same code to work on a local or remote instance of sbot. also note, that when a rpc function is called it's called with the context `{id: remote_id}` so that `this.id == remote_id` which makes it possible to know which peer called this function. That is used for implementing blocks for example. That's my understanding of it. It's up otherwise up to clients to keep track of their own id.

christianbundy replied:
Thanks for the info. I've just pushed a commit to fix this and make sure it's set as the remote ID. Please let me know if you'd like to see any other changes!

dominictarr commented:
I don't think it's a good idea to completely remove this option, because it adds another round trip to establishing the connection. You might not notice that in CLI, but if you are accepting an invite code from some tiny pub running on a rpi behind dynamic dns or through tor, or on satellite internet, then latency becomes quite significant. ssb replication is carefully designed to minimize round trips!

If this was the default if a manifest wasn't provided, that would be okay, though.

christianbundy commented:
Sweet, I'll change this PR to re-add `manifest` back as an option.

christianbundy commented:
Fixed! I've also gone through the files I've changed and tried to make their style more consistent. I specifically left the visual whitespace changes that seemed intentional.

christianbundy commented on test/index.js:
```diff
@@ -8,29 +8,56 @@ var ssbClient = require('../')
+    caps: { shs: shsCap }
+  })
+  server.on('multiserver:listening', () => {
+    ssbClient(keys, { port: 45451, manifest: server.manifest(), caps: { shs: shsCap } }, function (err, client) {
```
Done! Just pushed a commit to test the `{ foo: 'async' }` manifest.

christianbundy commented:
Thanks @soapdog!
